Enable all and any for bool tensors

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21033 Enable all and any for bool tensors**
* #21032 Enabled add, sum, mul and neg for bool tensor
* #21031 Enabled min, max, minall, maxall, cmin, cmax, cmaxValue, cminValue for bool tensors
* #21030 Added possibility to index scalars by bool masks

This PR is a part of a stack which will change result tensor type of comparison ops from uint8 to bool. As this change is rather big and a lot of prep work is needed, im breaking it into a stack.

Changes in this PR:
- Enable all() and any() for bool tensors.

Differential Revision: [D15530497](https://our.internmc.facebook.com/intern/diff/D15530497)

izdeby committed:
Enable all and any for bool tensors

colesbury reviewed (commented):
This needs a test for non-empty any/all

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "[WIP] Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "[WIP] Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -531,12 +531,12 @@ Tensor all(const Tensor& self) {
 
   Tensor result = at::empty({0}, self.options());
   auto iter = make_reduction(
-    "all", result, self, {}, false, at::ScalarType::Byte);
```
Good.

colesbury replied:
This is fine for now, but the explicit return type was intentional. numpy.all and numpy.any always return a `bool` array even for non-bool inputs.

ezyang reviewed (approved):
Approved with comments; I agree with Sam, non-empty case needs testing (or better yet, find where we are testing all/any previously and generalize that test code.)

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby commented:
@pytorchbot retest this please

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head

izdeby committed:
Update on "Enable all and any for bool tensors"

Enable all and any for bool tensors

gh-metadata: pytorch pytorch 21033 gh/izdeby/4/head
